src/SyntaxHighlighter/Cpp.cpp

```diff
@@ -0,0 +1,107 @@
+    "double|else|enum|extern|float|for|goto|if|int|long|register|return|short|signed|sizeof|static|"
+    "struct|switch|typedef|union|unsigned|void|volatile|while|)\\b";
+const char* const kCommentRegex =
+    "\\/\\/(?:[^\\r\\n\\\\]|\\\\(?:\\r\\n?|\\n|(?![\\r\\n])))*|\\/\\*[\\s\\S]*?\\*\\/";
+const char* const kOpenCommentRegex = "\\/\\*([^\\*]|[\\*]+[^\\/])*?$";  // Don't work with **/
+const char* const kEndCommentRegex = "[\\s\\S]*\\*\\/";
+const char* const kFunctionRegex = "[a-z_]\\w*(?=\\s*\\()";
+const char* const kNoEndCommentRegex = "([^\\*]|\\*+[^\\/\\*])*$";
+const char* const kMacroRegex = "(^\\s*)#\\s*[a-z_]\\S*";
+const char* const kIncludeFileRegex = "^(#\\s*include\\s*)<[^>]+>";
+const char* const kStringRegex = "\"([^\\\\\"]|\\\\.)*\"|\'[^\']*\'";
+const char* const kOpenStringRegex = "\"([^\\\\\"]|\\\\.)*\\\\$";
+const char* const kEndStringRegex = "([^\\\\\"]|\\\\.)*\"";
+const char* const kNoEndStringRegex = "([^\\\\\"]|\\\\.)*\\\\$";
+const char* const kUppercaseRegex = "[A-Z][\\w\\*\\&]*";
+}  // namespace
+}  // namespace C
+
+Cpp::Cpp() : QSyntaxHighlighter{static_cast<QObject*>(nullptr)} {
+  using PatternOption = QRegularExpression::PatternOption;
+  number_regex_ = QRegularExpression{C::kNumberRegex, PatternOption::CaseInsensitiveOption};
+  constant_regex_ = QRegularExpression{C::kConstantRegex};
+  keyword_regex_ = QRegularExpression{C::kKeywordRegex};
+  comment_regex_ = QRegularExpression{C::kCommentRegex};
+  open_comment_regex_ = QRegularExpression{C::kOpenCommentRegex};
+  end_comment_regex_ = QRegularExpression{C::kEndCommentRegex};
+  no_end_comment_regex_ = QRegularExpression{C::kNoEndCommentRegex};
+  function_regex_ = QRegularExpression{C::kFunctionRegex, PatternOption::CaseInsensitiveOption};
+  macro_regex_ = QRegularExpression{C::kMacroRegex};
+  include_file_regex_ = QRegularExpression{C::kIncludeFileRegex};
+  string_regex_ = QRegularExpression{C::kStringRegex};
+  open_string_regex_ = QRegularExpression{C::kOpenStringRegex};
+  end_string_regex_ = QRegularExpression{C::kEndStringRegex};
+  no_end_string_regex_ = QRegularExpression{C::kNoEndStringRegex};
+  uppercase_regex_ = QRegularExpression{C::kUppercaseRegex};
+}
+
+void Cpp::highlightBlock(const QString& code) {
+  const auto apply = [&code, this](
+                         QRegularExpression* expression, const QColor& color,
+                         C::HighlighterState new_state = C::HighlighterState::kInitialState) {
+    QTextCharFormat format{};
+    format.setForeground(color);
+
+    for (auto it = expression->globalMatch(code); it.hasNext();) {
+      const auto match = it.next();
+      setFormat(match.capturedStart(), match.capturedLength(), format);
+      setCurrentBlockState(new_state);
+    }
+  };
+
+  apply(&number_regex_, C::kBlue);
+  apply(&uppercase_regex_, C::kViolet);
+  apply(&keyword_regex_, C::kOrange);
+  apply(&include_file_regex_, C::kGreen);
```

Comment:
Ok, I'm really bad with reading regular expressions, so I'm sorry if I'm getting this wrong. 

It looks to me that at least the "include file" and "string" and "macro" regex work together in interesting ways here and that the order matters. 

"Include file regex" will turn green a string like '#include<header.h>'. "String regex" will turn green a string like '"this"' including '"header.h"'. And finally macro regex turns '#pragma', but also '#include', into yellow orange.

So, 

#include <file.h>
#include "header.h"

turn out the same, but due to different rules. 

Doing the "include file regex" after "macro regex" would also lead to a different result (example: #include with <> and "" would look different), right? That may be fine and intended, but I would add a comment to explain that.

Reply:
There are two different things
1- Yes, I tried to follow the same rules than X86Assembly where we process line by line and apply the rules one on top of another. So, the order matters. If we have **"Hi world, I'm an int variable"**, first **"int"** is painted and after this is overpainted by the comment thing". This leads to one problem, what comes first? Comment or String?
_" /* Hi, I'm an string */. ""_
_/*  "Hi, I'm a comment" */_
Is impossible to solve with the approach, but as they are really rare, we are continuing with our approach. If not, a more complex process is needed. But if you feel that it is important, I can change it.
2- The fact with macro and string was intended. But now, thinking about it better, I'm realizing that I can avoid that without much effort. Will change it.

Reply:
As far as I can tell, the approach is fine for our purposes. I would suggest to document it a bit, at least a comment in the code that the order matters and shouldn't be changed unless one knows what they are doing.

Reply:
Added a few comments in the regex and also when we're applying them. My goal is include at the first the "heuristic ones", where we can modify them following C++/Google conventions and to make the code feels a little bit nicer.